First attempt at parse-uri and build-uri functions

ndw opened the pull request:
Close #72 

This is a first attempt to resolve issue 72 by describing `parse-uri` and `build-uri` functions.

I decided to try an approach that leverages the ABNF defined in RFC 3986. Attempting to redefine the rules just opens up the possibility that the rules we write will differ from the rules in RFC 3986 which would be wrong.

I added an options map and specified a "details" option that will further parse the path and query components. This gives the caller easy access to the percent-decoded forms of those elements without making the parse function irreversable.

Notes:

1. More examples are needed
2. I tried to use `eg` in the results so that they'd be formatted nicely, but it didn't work. Is there any markup for verbatim text in examples?
3. There's an open question about whether non-hierarchical URIs can have fragment identifiers or queries. I think not, but I need to review the RFC a little more carefully.

ndw commented:
Formatted versions of the specifications with this PR applied [are available](https://qt4cg.org/#pr-215)

ndw commented:
I started with a vision of parsing RFC 3986 (absolute) URIs. I think Mike makes a very good point that one reason to parse a URI is to find out things like, is it absolute or relative? The trouble is, once you start accepting those strings, you quickly find yourself up against special cases.

Experience has shown, for example, that `file:` URIs on Windows are especially problematic. One often encounters  `file:\\c:\path` which isn't even really a URI, though it's "obvious" what was intended. Constraining ourselves to actual URIs, we still get awkward things like `file://c:/path` and `file:////unc/path`. (There is no specification for the `file:` URI scheme, so we can't appeal to anyone for help.)

How far down the road to [p:urify()](https://spec.xproc.org/3.0/xproc/#f.urify) do we want to go?

I'm tempted to add things like:

> If a string begins with a single letter followed by a colon, assume it's a file: URI on Windows.

and 

> Convert all backslashes in the string into forward slashes, before analyzing the string

I think we can get something useful and interoperable this way, but it'll be a little ad hoc of necessity.

michaelhkay commented:
Just been reading p:urify(). At least it reads like a spec. Everything else I've seen reads like arm-waving.

ndw commented:
1. I'd be happy to have `fn:urify()`, it doesn't do exactly what parse and build do.
2. The implementation of `fn:urify()` is specifically different on Windows versus other environments. I don't know if we want to do that or not.

gimsieke commented:
The `p:un` would be lost on `fn:urify` though.

ndw commented:
I’ve done a pretty substantial reworking of *parse-uri()* based on my attempts to address the feedback here.

The apparently spurious commits caused by rebasing on master were necessary in order to get some stylesheet updates into the build that will make the formatted version much easier to read!

Mike wrote:

> the function is described as context-independent, but then it says it expands a relative URI using the static base URI.
> 
> I'm not sure expanding a relative URI is a good thing. One thing you want to know about the URI is whether it's relative or absolute.

Right. Okay. Let’s see what happens if we just make it operate on the string passed in. The new prose is a bit procedural, but that’s (explicitly) about exposition and not implementation.

> There are no error conditions defined, which seems a little surprising.

I decided that it shouldn’t raise errors. This is especially true if it’s just going to attempt to parse random strings.

> The ABNF that I'm looking at uses different names for the URI parts: iuserinfo rather then userinfo, etc. So we need to define the mapping.

`iuserinfo`? What spec are you looking at? I’m reading [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986).

> It's not clear to me whether absent components should appear in the result map using a value of "", or (), or should be omitted.

I’ve currently opted for making them appear in the map as `()`. I could be persuaded that absent values should be left out of the map. In fact, in the examples, I’ve left them out simplicity, which is maybe kind of a clue.

> I would be inclined to do the details=true() option unconditionally.

Yes.

> Should allow ";" as an alternative to "&" between query parts? I know it's not widely used but it's there and is often more convenient than "&" in an XML context.

It’s not standard. But I added it as an option. I added a path separator option as well.

> The expansion of the query part as a sequence of maps preserves all the information, but isn't as easy to use as, for example, returning a single map in which multiple entries for the same keyword are combined into a sequence (So ?x=3&x=4 becomes map{'x', (3,4)} rather than (map{'x':3}, map{'x':4}).

I’ve replaced that with an array. You have to parse on the equal sign yourself.

> "Escaped" should probably be "percent-encoded".

That text is gone, I think.

Arithmeticus commented:
Thanks for all this work, Norm. The variety of examples are very useful. Could you also include a tag URI example?
https://tools.ietf.org/html/rfc4151

e.g., 
tag:textalign.net,2015:ns
tag:jan@example.com,1999-01-31:my-uri

ndw commented:
Sure, @Arithmeticus. FYI:
```
map {
    "uri": "tag:textalign.net,2015:ns",
    "scheme": "tag",
    "path": "textalign.net,2015:ns",
    "path-segments": [ "textalign.net,2015:ns" ]
  }
```
and
```
map {
    "uri": "tag:jan@example.com,1999-01-31:my-uri"
    "scheme": "tag",
    "path": "jan@example.com,1999-01-31:my-uri",
    "path-segments": [ "jan@example.com,1999-01-31:my-uri" ],
}
```

ndw commented:
Updated per comments from [meeting 009](https://qt4cg.org/meeting/minutes/2022/11-01.html#pr-parse-uri):

1. Added prose describing how parse-uri and resolve-uri are related.
2. Reviewed RFC 3986 wrt relative URIs and reworded the introduction.
3. Made the return-type a record, mad the argument type in `build-uri` a record, and updated the stylesheets to support a more readable rendering if record types as arguments and return types
4. Added an error test for parse-uri; added a non-error test demonstrating the use of `query-separator`
5. Made the `query-segments` an array of maps
6. Redrafted the prose for `build-uri`.
7.

ndw commented:
Accepted at [meeting 011, 15 November 2022](https://qt4cg.org/meeting/minutes/2022/11-15.html)
